src/dma/mod.rs

```diff
@@ -0,0 +1,148 @@
+#![allow(dead_code)]
+
+#[repr(u8)]
+pub enum c_void {
+    __variant1,
+    __variant2,
+}
+
+pub const DMAREQ_TIMER0_UFOF: u32 = ((24 << 16) + 0);
+
+pub type FuncPtr = extern fn(channel: u32, primary: bool, user: u32);
+
+use core::intrinsics::transmute;
+use core::default::Default;
+use core::option::Option;
+use core::slice::SliceExt;
+use core::ptr;
+
+pub struct DMA { channel: u32 }
+
+impl DMA {
+    pub fn channel0() -> DMA {
+        DMA { channel: 0 }
+    }
+
+    pub fn configure_channel(&self, cfg: &CfgChannel) {
+        unsafe { DMA_CfgChannel(self.channel, cfg) }
+    }
+
+    pub fn configure_descriptor(&self, primary: bool, cfg: &CfgDescriptor) {
+        unsafe { DMA_CfgDescr(self.channel, primary, cfg) }
+    }
+
+    pub fn activate_auto<T>(&self, primary: bool, dst: &mut[T], src: &mut[T]) {
+        unsafe {
+
+            let (dst_n, src_n) = (dst.len(), src.len());
+
+            let n = if dst_n < src_n { dst_n } else { src_n };
+            
+            DMA_ActivateAuto(
+                self.channel,
+                primary,
+                transmute(dst.as_ptr()),
+                transmute(src.as_ptr()),
+                n - 1
+            );
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Copy)]
+pub struct Init {
+    pub hprot: u8,
+    pub control_block: &'static Descriptor,
+}
+
+#[repr(C)]
+#[derive(Copy)]
+pub struct CfgChannel {
+    pub high_pri: bool,
+    pub enable_int: bool,
+    pub select: u32,
+    pub cb: &'static CB,
+}
+
+#[repr(C)]
+#[derive(Copy)]
+pub struct CfgDescriptor {
+    pub dst_inc: DataInc,
+    pub src_inc: DataInc,
+    pub size: DataSize,
+    pub arb_rate: ArbiterConfig,
+    pub hprot: u8,
+}
+
+#[repr(C)]
+#[derive(Copy)]
+pub struct Descriptor {
+    pub SRCEND: u32,
+    pub DSTEND: u32,
+    pub CTRL: u32,
```

Comment:
Generally, I think we should strive towards getting rid of all warnings. 
I think we should consider adding ` #[allow(non_camel_case_types)]` for types that generate this warning, if we want them not to have snake_case.